cmd/webhook/admission/admission.go

```diff
@@ -0,0 +1,218 @@
+	}, labels)
+	approvedRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "successful_admissions",
+		Help:      "Total number of requests successfully processed by this admitter",
+	}, labels)
+	admissionDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "admission_duration",
+		Help:      "Duration of admission calls",
+		Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, .75, 1, 1.25, 1.5, 2, 2.5, 5, 10},
+	}, labels)
+)
+
+type Admitter interface {
+	Name() string
+	Admit(req *admissionsv1.AdmissionRequest) (*admissionsv1.AdmissionResponse, error)
+}
+
+type RouteGroupAdmitter struct {
+}
+
+func init() {
+	prometheus.MustRegister(totalRequests, invalidRequests, rejectedRequests, approvedRequests, admissionDuration)
+}
+
+func (r RouteGroupAdmitter) Name() string {
+	return "routegroup"
+}
+
+func (r RouteGroupAdmitter) Admit(req *admissionsv1.AdmissionRequest) (*admissionsv1.AdmissionResponse, error) {
+	rgItem := definitions.RouteGroupItem{}
+	err := json.Unmarshal(req.Object.Raw, &rgItem)
+	if err != nil {
+		emsg := fmt.Errorf("could not parse RouteGroup, %w", err)
+		log.Error(emsg)
+		return &admissionsv1.AdmissionResponse{
+			UID:     req.UID,
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: emsg.Error(),
+			},
+		}, nil
+	}
+
+	err = definitions.ValidateRouteGroup(&rgItem)
+	if err != nil {
+		emsg := fmt.Errorf("could not validate RouteGroup, %w", err)
+		log.Error(emsg)
+		return &admissionsv1.AdmissionResponse{
+			UID:     req.UID,
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: emsg.Error(),
+			},
+		}, nil
+	}
+
+	return &admissionsv1.AdmissionResponse{
+		UID:     req.UID,
+		Allowed: true,
+	}, nil
+}
+
+func Handler(admitter Admitter) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		admitterName := admitter.Name()
+		totalRequests.WithLabelValues(admitterName).Inc()
+
+		if r.Method != "POST" || r.Header.Get("Content-Type") != "application/json" {
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			w.WriteHeader(http.StatusBadRequest)
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Errorf("Unable to read request: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			return
+		}
+
+		review := admissionsv1.AdmissionReview{}
+		err = json.Unmarshal(body, &review)
+		if err != nil {
+			log.Errorf("Unable to parse request: %v", err)
+			w.WriteHeader(http.StatusBadRequest)
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			return
+		}
+
+		operationInfo := fmt.Sprintf(
+			"%s %s %s/%s",
+			review.Request.Operation,
+			review.Request.Kind,
+			review.Request.Namespace,
+			extractName(review.Request),
+		)
+
+		gvr := review.Request.Resource
```

Comment:
the acronym `gvr` is not clear to me what it means

Reply:
It's Group Version Resource https://godoc.org/k8s.io/apimachinery/pkg/runtime/schema#GroupVersionResource happy to make it more clear by a comment or rename

Reply:
thx! TIL :+1: